### PR TITLE
Added replacement pcb for the matrix cain keyboard

### PR DESCRIPTION
**Pull request by yulei**
added replacement pcb for the matrix cain keyboard

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [ x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x ] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/matrix/cain/cain.c [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+#include "cain.h"
+
+void matrix_init_user(void)
+{
+	setPinOutput(NUM_PIN);
+	setPinOutput(CAPS_PIN);
+	setPinOutput(SCROLL_PIN);
+}
```
Comment (zvecr) [inline keyboards/matrix/cain/cain.c:27]:
```suggestion
void matrix_init_kb(void)
{
	setPinOutput(NUM_PIN);
	setPinOutput(CAPS_PIN);
	setPinOutput(SCROLL_PIN);
	
	matrix_init_user();
}
```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/matrix/cain/rules.mk [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+#
+BOOTMAGIC_ENABLE = yes		# Virtual DIP switch configuration
+MOUSEKEY_ENABLE = yes		# Mouse keys
+EXTRAKEY_ENABLE = yes		# Audio control and System control
+CONSOLE_ENABLE = no			# Console for debug
+COMMAND_ENABLE = no			# Commands for debug and configuration
+SLEEP_LED_ENABLE = no		# Breathing sleep LED during USB suspend
+NKRO_ENABLE = yes			# USB Nkey Rollover
+BACKLIGHT_ENABLE = no		# Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
```
Comment (fauxpark) [inline keyboards/matrix/cain/rules.mk:18]:
```suggestion
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = full     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/matrix/cain/cain.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+#include "quantum.h"
+
+#define LAYOUT_default( \
+    k000,       k001, k002, k003, k004,   k005, k006, k007, k008,     k009, k010, k011, k012,                         k013, k014, k015, \
+    k100, k101, k102, k103, k104, k105, k106, k107, k108, k109, k110, k111, k112,       k113, k114, k115, k116, k117, k118, k119, k120, \
+    k200,    k201, k202, k203, k204, k205, k206, k207, k208, k209, k210, k211, k212,    k213, k214, k215, k216, k217, k218, k219, k220, \
+    k300,       k301, k302, k303, k304, k305, k306, k307, k308, k309, k310, k311,       k312, k313, k314, k315, k316, \
+    k400,         k401, k402, k403, k404, k405, k406, k407, k408, k409, k410,           k411, k412, k413, k414, k415,       k416, \
+    k500,   k501,     k502,             k503,                    k504,   k505,   k506,  k507, k508, k509, k510,       k511, k512, k513 \
```
Comment (noroadsleft) [inline keyboards/matrix/cain/cain.h:29]:
```suggestion
#define LAYOUT_all( \
    k000,       k001, k002, k003, k004,   k005, k006, k007, k008,     k009, k010, k011, k012,                             k013, k014, k015, \
    k100, k101, k102, k103, k104, k105, k106, k107, k108, k109, k110, k111, k112,       k113,   k114, k115, k116, k117,   k118, k119, k120, \
    k200,    k201, k202, k203, k204, k205, k206, k207, k208, k209, k210, k211, k212,    k213,   k214, k215, k216, k217,   k218, k219, k220, \
    k300,       k301, k302, k303, k304, k305, k306, k307, k308, k309, k310, k311,       k312,   k313, k314, k315, k316, \
    k400,         k401, k402, k403, k404, k405, k406, k407, k408, k409, k410,           k411,   k412, k413, k414, k415,         k416, \
    k500,     k501,   k502,                   k503,                   k504,   k505,     k506,   k507, k508, k509, k510,   k511, k512, k513 \
```

File: keyboards/matrix/cain/cain.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+) { \
+  { k000, k001, k002, k003, k004, k005, k006, k007, k008, k009, k010, k011}, \
+  { k100, k101, k102, k103, k104, k105, k106, k107, k108, k109, k110, k111}, \
+  { k200, k201, k202, k203, k204, k205, k206, k207, k208, k209, k210, k211}, \
+  { k300, k301, k302, k303, k304, k305, k306, k307, k308, k309, k310, k311}, \
+  { k400, k401, k402, k403, k404, k405, k406, k407, k408, k409, k410, k411}, \
+  { k012, k113, k112, k212, k213, k312, k412, k413, k414, k415, k316, k315}, \
+  { k114, k115, k116, k117, k217, k216, k215, k214, k313, k314, k416}, \
```
Comment (noroadsleft) [inline keyboards/matrix/cain/cain.h:37]:
```suggestion
  { k114, k115, k116, k117, k217, k216, k215, k214, k313, k314, k416, KC_NO}, \
```

File: keyboards/matrix/cain/cain.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  { k013, k014, k015, k120, k119, k118, k218, k219, k220, k511, k512, k513}, \
+  { k500, k501, k502, k503, k504, k505, k506, k507, k508, k509, k510} \
```
Comment (noroadsleft) [inline keyboards/matrix/cain/cain.h:39]:
```suggestion
  { k500, k501, k502, k503, k504, k505, k506, k507, k508, k509, k510, KC_NO} \
```

File: keyboards/matrix/cain/info.json [hunk 6]
```diff
@@ -0,0 +1,12 @@
+{
+    "keyboard_name": "Matrix Cain Keyboard",
+    "url": "",
+    "maintainer": "qmk",
+    "width": 23,
+    "height": 6.5,
+    "layouts": {
+        "LAYOUT_default": {
```
Comment (noroadsleft) [inline keyboards/matrix/cain/info.json:8]:
```suggestion
        "LAYOUT_all": {
```

File: keyboards/matrix/cain/keymaps/default/keymap.c [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+	[0]=LAYOUT_default(
```
Comment (noroadsleft) [inline keyboards/matrix/cain/keymaps/default/keymap.c:23]:
```suggestion
	[0]=LAYOUT_all(
```

File: keyboards/matrix/cain/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+		KC_ESC, KC_F1, KC_F2, KC_F3, KC_F4, KC_F5, KC_F6, KC_F7, KC_F8, KC_F9, KC_F10, KC_F11, KC_F12,                                     KC_PSCR, KC_SLCK, LT(1,KC_PAUS),
+
+		KC_GRV, KC_1, KC_2, KC_3, KC_4, KC_5, KC_6, KC_7, KC_8, KC_9, KC_0, KC_MINS, KC_EQL,  KC_BSPC, KC_NLCK, KC_PSLS, KC_PAST, KC_PMNS,  KC_INS, KC_HOME, KC_PGUP,
+		KC_TAB, KC_Q, KC_W, KC_E, KC_R, KC_T, KC_Y, KC_U, KC_I, KC_O, KC_P, KC_LBRC, KC_RBRC, KC_BSLS,   KC_P7,   KC_P8,   KC_P9, KC_PPLS,  KC_DEL,  KC_END, KC_PGDN,
+		KC_CAPS,   KC_A, KC_S, KC_D, KC_F, KC_G, KC_H, KC_J, KC_K, KC_L, KC_SCLN, KC_QUOT,     KC_ENT,   KC_P4,   KC_P5,   KC_P6, KC_PPLS,
+		KC_LSFT,     KC_Z, KC_X, KC_C, KC_V, KC_B, KC_N, KC_M, KC_COMM, KC_DOT, KC_SLSH,      KC_RSFT,   KC_P1,   KC_P2,   KC_P3, KC_PENT,            KC_UP,
+		KC_LCTL, KC_LGUI, KC_LALT,                 KC_SPC,                    KC_RALT, MO(1), KC_RCTL,   KC_P0,   KC_P0, KC_PDOT, KC_PENT, KC_LEFT, KC_DOWN, KC_RGHT),
+
+	[1]=LAYOUT_default(
```
Comment (noroadsleft) [inline keyboards/matrix/cain/keymaps/default/keymap.c:32]:
```suggestion
	[1]=LAYOUT_all(
```

File: keyboards/matrix/cain/read.md [hunk 9]
```diff
@@ -0,0 +1,15 @@
+# Matrix Cain keyboard
+
+This was the replacement PCB for the Matrix Cain keyboard. 
+
+This firmware DO NOT support the original PCB shipped with the Cain keybaord.
```
Comment (noroadsleft) [inline keyboards/matrix/cain/read.md:5]:
```suggestion
This firmware **DOES NOT** support the original PCB shipped with the Cain keyboard.
```

File: keyboards/matrix/cain/read.md [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+* Keyboard Maintainer: [astro](https://github.com/yulei)
+* Hardware Supported: Matrix cain keyboard
```
Comment (noroadsleft) [inline keyboards/matrix/cain/read.md:8]:
```suggestion
* Hardware Supported: Matrix Cain keyboard
```

File: keyboards/matrix/cain/read.md [hunk 11]
```diff
@@ -0,0 +1,15 @@
+# Matrix Cain keyboard
```
Comment (noroadsleft) [inline keyboards/matrix/cain/read.md:1]: This file should be named `readme.md`.

```
git mv -- keyboards/matrix/cain/read.md keyboards/matrix/cain/readme.md
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]
GitHub tip: You can apply multiple suggestions to a single commit by using the [Files Changed tab](https://github.com/qmk/qmk_firmware/pull/10436/files).

Also, as this code doesn't support the original Matrix Cain PCB, I'd much prefer if the directory name was something other than `cain` – something like `cain_redux`, maybe, though I'm open to suggestions here.

**Comment by yulei** [comment]
> GitHub tip: You can apply multiple suggestions to a single commit by using the [Files Changed tab](https://github.com/qmk/qmk_firmware/pull/10436/files).
> 
> Also, as this code doesn't support the original Matrix Cain PCB, I'd much prefer if the directory name was something other than `cain` – something like `cain_redux`, maybe, though I'm open to suggestions here.

selected new name as `cain+re` :)

File: keyboards/matrix/cain_re/readme.md [hunk 12]
```diff
@@ -0,0 +1,15 @@
+# Matrix Cain keyboard
```
Comment (noroadsleft) [inline keyboards/matrix/cain_re/readme.md:1]:
```suggestion
# Matrix Cain replacement PCB
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]
Tiny change here. Everything else looks good to me.

**Review by noroadsleft — APPROVED** [review]

**Review by zvecr — APPROVED** [review]

**Comment by zvecr** [comment]
Thanks!

**Pull request merged**